storage/shard.go

```diff
@@ -762,21 +814,36 @@ func (s *dbShard) Write(
 		// as the commit log need to use the reference without the
 		// overhead of ownership tracking. This makes taking a ref here safe.
 		commitLogSeriesID = entry.series.ID()
+		commitLogSeriesTags = entry.series.Tags()
 		commitLogSeriesUniqueIndex = entry.index
-		entry.decrementReaderWriterCount()
 		if err != nil {
+			entry.decrementReaderWriterCount()
 			return err
 		}
+		needsIndex := reverseIndexWriteFn != nil && entry.needsIndexUpdate(timestamp)
+		if needsIndex {
+			// NB(prateek): we retain the ref count on the entry while the indexing is pending,
+			// the callback executed on the entry once the indexing is completed releases this
+			// reference.
+			reverseIndexWriteFn(entry.series.ID(), entry.series.Tags(), entry)
+		} else {
+			entry.decrementReaderWriterCount()
+		}
```

Comment:
To simplify the code a little above, this looks slightly cleaner:
```go
		indexSeries := reverseIndexWriteFn != nil && entry.needsIndexUpdate(timestamp)
		if err == nil && indexSeries {
			// OnIndexFinalize will decrement the reader/writer
			reverseIndexWriteFn(entry.series.ID(), entry.series.Tags(), entry)
		} else {
			entry.decrementReaderWriterCount()
		}
		if err != nil {
			return err
		}
```

Reply:
Actually on second thoughts, why don't we leave the code like it was but do a second increment if we need to index, similar to the async batch insert?  Seems better perhaps.  Then it would also be more clean if we also added the synchronous method name, i.e. `entry.onIndexPrepare()` which would increment the reader/writer count.

Reply:
sure, will do

Reply:
done

Reply:
done